include/swift/AST/Expr.h

```diff
@@ -553,6 +623,81 @@ class alignas(8) Expr {
   }
 };
 
+/// Helper class to capture trailing call argument labels and related
+/// information, for expression nodes that involve argument labels, trailing
+/// closures, etc.
+template<typename Derived>
+class TrailingCallArguments
+    : private llvm::TrailingObjects<Derived, Identifier, SourceLoc> {
+  using TrailingObjects = llvm::TrailingObjects<Derived, Identifier, SourceLoc>;
+  friend TrailingObjects;
+
+  Derived &asDerived() {
+    return *static_cast<Derived *>(this);
+  }
+
+  const Derived &asDerived() const {
+    return *static_cast<const Derived *>(this);
+  }
+
+  size_t numTrailingObjects(
+      typename TrailingObjects::template OverloadToken<Identifier>) const {
+    return asDerived().getNumArguments();
+  }
+
+  size_t numTrailingObjects(
+      typename TrailingObjects::template OverloadToken<SourceLoc>) const {
```

Comment:
Hi @DougGregor, sorry to bring this up from the past.

I'm trying (and not succeeding) to build swiftAST using MSVC. I've managed to get most things to work, fixing errors and MSVC bugs (my god, there are a lot of them).

I have one final hurdle I've been unable to cross. I get an error saying the following (on this line):
```
Error	C2751 'llvm::TrailingObjects<BaseTy,TrailingTys...>::operator OverloadToken': the name of a function parameter cannot be qualified (compiling source file lib\AST\ASTNode.cpp
```

Since you added the code, I just wanted to inquire as to what you were trying to do here. Perhaps MSVC is being buggy, or maybe Clang was letting this pass when it shouldn't have. Maybe this code is invalid according to the C++ Specification.
 If you have any suggestions, I'd really appreciate it!